# Graph neural networks



In [1]:
import os

import numpy as np
import pandas as pd
from sklearn.metrics import mean_squared_error

import torch
import torch.nn as nn
import torch.nn.functional as F

from torch_geometric.data import Data
from torch_geometric.data.separate import separate
from torch_geometric.datasets import Planetoid, ZINC
from torch_geometric.loader import DataLoader, LinkNeighborLoader
from torch_geometric.utils import scatter
from torch_geometric.nn import GCNConv, GATConv, TransformerConv
from torch_geometric.utils import to_dense_adj

import plotly.express as px

## Data for node prediction

A now classic dataset for GNNs (one whose use is also discouraged within some circles as it is very easy to overfit on), Cora is a nice small dataset to start looking at GNNs. There are many variations of the Cora dataset originally presented in "Automating the Construction of Internet Portals with Machine Learning" by McCallum et al. (https://link.springer.com/article/10.1023/A:1009953814988).

We will use the Cora dataset variant as presented in “FastGCN: Fast Learning with Graph Convolutional Networks via Importance Sampling” (https://arxiv.org/abs/1801.10247). It describes a citation network of 2708 papers and our task is classify each paper into one of 7 different categories.

In [2]:
# Load dataset:
data_dir = "./cora_data_full"
try: 
    os.mkdir(data_dir)
except FileExistsError:
    print(data_dir, 'exists')
    
dataset = Planetoid(data_dir, 'Cora', split="full")

Processing...
Done!


Split into training, validation and test nodes.

In [6]:
# train-test split
train_x = dataset.x[dataset.train_mask]
train_y = dataset.y[dataset.train_mask]

valid_x = dataset.x[dataset.val_mask]
valid_y = dataset.y[dataset.val_mask]

test_x = dataset.x[dataset.test_mask]
test_y = dataset.y[dataset.test_mask]

Describe the training data. 

In [15]:
dataset.x.shape, train_x.shape, train_y.shape, np.unique(train_y)

(torch.Size([2708, 1433]),
 torch.Size([1208, 1433]),
 torch.Size([1208]),
 array([0, 1, 2, 3, 4, 5, 6]))

The dataset contains 2708 paper, each with up to 1433 features. Each paper has a label/category from 0 to 6.

## Baseline: feed-forward neural network

In [17]:
class FFNN(nn.Module):
    """A simple feed forward neural network with no hidden layers

    Args:
        input_dim (int): Dimensionality of the input feature vectors
        output_dim (int): Dimensionality of the output softmax distribution
    """
    def __init__(self, input_dim, output_dim):
        super().__init__()
        self.layer_1 = nn.Linear(input_dim, output_dim)

    def forward(self, x):
        x = self.layer_1(x)
        return F.log_softmax(x, dim=1)

In [18]:
n_epochs = 100
model = FFNN(input_dim=train_x.shape[-1], output_dim=7)
optimiser = torch.optim.Adam(model.parameters(), lr=0.001)

val_accuracy = {key: 0 for key in range(n_epochs)}

for epoch in range(n_epochs):
    model.train()
    optimiser.zero_grad()
    y_hat = model(train_x)
    loss = F.cross_entropy(y_hat, train_y)
    loss.backward()
    optimiser.step()
    
    model.eval()
    y_hat = model(valid_x)
    y_hat = y_hat.data.max(1)[1]
    num_correct = y_hat.eq(valid_y.data).sum()
    num_total = len(valid_y)
    val_accuracy[epoch] = 100.0 * (num_correct.detach().numpy()/num_total)

In [19]:
fig = px.scatter(x=range(n_epochs), y=np.array(list(val_accuracy.values())))

fig.update_layout(template="plotly_white", 
                xaxis_title="epoch",
                yaxis_title="validation accuracy (in %)")

fig.show()

## Do it yourself graph neural network

The generic equation for convolutional GNN is
$$
\mathbf{h_i} = \phi \big(\mathbf{x_i}, \oplus_{j \in \mathcal{N}_i} c_{i,j} \psi (\mathbf{x_j}) \big).
$$
Rewrite this in terms of adjacency matrix $\mathbf{A}$ and implement it.

We will use the mean sum as the aggregated function, and let the convolution be

$$
    h_i = \phi \left( \sum_{j \in \mathcal{N}_i} A_{i,j}X_{j}\Theta_{j} \right)
$$

In [78]:
class GCLayer(nn.Module):
    """Graph convolution layer

    Args:
        input_dim (int): Dimensionality of the input feature vectors
        output_dim (int): Dimensionality of the output softmax distribution
        A (torch.Tensor): 2-D adjacency matrix
    """
    def __init__(self, input_dim, output_dim, A):
        super(GCLayer, self).__init__()
        self.input_dim = input_dim
        self.output_dim = output_dim
        self.A = A

        # Let the weight matrix be stored in the lienar layer
        self.linear = nn.Linear(input_dim, output_dim)
        self.theta = nn.Parameter(torch.rand(input_dim, input_dim))  # Initialize weights between [0, 1]

        self.activation = nn.ReLU()
        
    def forward(self, x):
        ### TODO
        
        x = self.A @ x @ self.theta
        # print(x.shape)
        # mat1 and mat2 shapes cannot be multiplied (2708x2708 and 1208x1433)
        x = self.activation(self.linear(x))
        # print(f"conv layer output shape: {x.shape}")
        return x


In [79]:
adj.shape

torch.Size([2708, 2708])

In [80]:
class GNN(nn.Module):
    """GNN model

    Args:
        input_dim (int): Dimensionality of the input feature vectors
        output_dim (int): Dimensionality of the output softmax distribution
        A (torch.Tensor): 2-D adjacency matrix
    """
    def __init__(self, input_dim, output_dim, A):
        super().__init__()
        self.input_dim = input_dim
        self.output_dim = output_dim
        self.layer = GCLayer(input_dim, output_dim, A)

    def forward(self, x):
        # print(x.shape)
        x = self.layer(x)
        # print(x.shape)
        return F.log_softmax(x, dim=1)

In [81]:
# adjacency matrix
adj = to_dense_adj(dataset.edge_index)[0]

In [82]:
dataset.train_mask

tensor([ True,  True,  True,  ..., False, False, False])

Trin the your graph neural network and plot the validation accuracies. 

In [83]:
n_epochs = 100
input_dim = dataset.x.shape[0]
output_dim = dataset.num_classes


model = GNN(input_dim=input_dim, output_dim=output_dim, A=adj)
optimiser = torch.optim.Adam(model.parameters(), lr=0.001)

val_accuracy = {key: 0 for key in range(n_epochs)}

for epoch in range(n_epochs):
    model.train()
    optimiser.zero_grad()
    y_hat = model(torch.eye(input_dim))
    loss = F.cross_entropy(y_hat[dataset.train_mask], train_y)
    loss.backward()
    optimiser.step()
    
#     model.eval()
#     valid_acc = sum(y_hat[dataset.valid_mask].max(1)[1] == valid_y)/nValid

#     if (epoch + 1) % 10 == 0:
#         print(f'Epoch [{epoch+1}/{num_epochs}], Tr Loss: {loss.item():.4f}, Vl Acc: {valid_acc.item():.4f}')

# test_acc = sum(output[test_idx].max(1)[1] == labels[test_idx])/nTest
# print('Test acc: %.2f'%(test_acc))

## Do it yourself graph neural network - normalization

The generic equation for convolutional GNN is
$$
\mathbf{h_i} = \phi \big(\mathbf{x_i}, \oplus_{j \in \mathcal{N}_i} c_{i,j} \psi (\mathbf{x_j}) \big).
$$

Specific implementations often differ slightly from this form. For example Kipf and Welling's "Semi-Supervised Classification with Graph Convolutional Networks" (https://arxiv.org/abs/1609.02907) employs a symmetric normalisation for the convolution coefficients with a re-normalisation to tackle exploding parameters. Here

$$
\mathbf{H} = \sigma \big( \mathbf{\tilde{D}}^{-\frac{1}{2}} \mathbf{\tilde{A}} \mathbf{\tilde{D}}^{-\frac{1}{2}} \mathbf{X} \mathbf{W} \big),
$$

where $\mathbf{\tilde{A}} = \mathbf{A} + \mathbf{I}$ and $\mathbf{\tilde{D}}$ is the degree matrix of $\mathbf{\tilde{A}}$.

Is the behaviour of the generic, convolutional GNN layer the same as Kipf and Welling's GCN? Implement this version.  How does the performance compare to the generic version above?

## Standard graph neural network layers

Have a look at some standard implementations of graph neural network layers, for example `GCNConv`, `GATConv` and `TransformerConv`. 

Note that these do not use the adjacency matrix, but the edge list (`data.edge_index`). 

Implement the models and show their validation accuracies during training. 

## Data for graph-level prediction

For this, we will use the ZINC dataset for graph-regression. It contains about 12 000 molecular graphs with up to 38 nodes each and the task is to predict for each molecule the solubility (a scalar number).

In [ ]:
# Load dataset:
data_dir = "./zinc_data"
try: 
    os.mkdir(data_dir)
except FileExistsError:
    print(data_dir, 'exists')

train_zinc_dataset = ZINC(root=data_dir, split='train', subset=True)
val_zinc_dataset = ZINC(root=data_dir, split='val', subset=True)
test_zinc_dataset = ZINC(root=data_dir, split='test', subset=True)

Describe this dataset. 

### GNN for graph-level prediction

Write a GNN for graph-level prediction and use it to predict the molecule solubility. 

Note: Using batch-sizes larger than `1` requires that you aggregate over the individual graphs, not the entire batch. To do this, I have used the following `extract`-function. 

In [ ]:
def extract(x, idx, slices):
    """ Extract x for graph with index idx
    Args: 
        x: data
        idx: index of graph in batch
        slices: indices where graphs start and end, e.g. batch.ptr
    """
    start, end = int(slices[idx]), int(slices[idx + 1])
    return x.narrow(0, start, end-start).sum(axis=0).reshape(1, -1)